### ISSUE TITLE
hangup_media not called when destroying session

### ISSUE THREAD

**Issue by douglascaetano** [meetecho/janus-gateway#1149]
When the browser client calls `janus.destroy()` while receiving a stream (e.g. when pressing "Stop" button on any demo), the current session is destroyed and, on Janus side, the `destroy_session` callback is called. I have expected, though, that `hangup_media` callback would be called before `destroy_session` callback, because the RTCPeerConnection would first be closed and only then the whole session would be destroyed. Is this the correct behavior?

**Comment by lminiero** [comment]
The way the core works right now, we only send events related to a handle when the resource exists. A "destroy" removes all handles as part of the process (which is what triggers PeerConnections to be closed too), and so when it comes at notifying about stuff happening there all events are discarded since the handle is no more. I guess that the callback could be invoked by `janus.js` directly, as it has knowledge of existing handles and the ones with a PeerConnection object.

**Comment by douglascaetano** [comment]
I see. I expected that `hangup_media` would be called always because then I wouldn't have to check in `destroy_session` for things that should have been freed or cleaned by `hangup_media`.

Personally, I think this should be the core behavior, i.e. doing all the "normal" steps on such situations, so plugins could trust more on the core's behavior and avoid doing repeated work. By "normal" I mean all the steps done in a common situation with a client starting a session, starting a stream, then stopping the stream and, finally, closing the session. That would be translated to calls, in sequence, to `create_session`, `setup_media`, `hangup_media` and `destroy_session` callbacks in every situation, even when the browser suddenly destroys the session like in my case. This trustfulness can greatly help in issues like memory leaks or use-after-free situations.

What you think? Would it be practical or applicable?

(By the way, I should have made myself clear that I'm making a custom plugin, so that's why I'm bringing this issue up.)

**Comment by lminiero** [comment]
Oh so you meant the callbacks in the plugin, not in JavaScript. Then you don't have to expect a `hangup_media` if you got a `destroy_session`: `hangup_media` is only called when the PeerConnection is closed due to external reasons, e.g., ICE failure or DTLS alert. When you get a `destroy_session`, you should invoke `hangup_media` internally to clean up resources. Have a look at how the stock plugins do that as an example.

Closing as not an issue.

**Comment by douglascaetano** [comment]
> Oh so you meant the callbacks in the plugin, not in JavaScript.

Yes, sorry. I should have made my question more clear.

> When you get a `destroy_session`, you should invoke `hangup_media` internally to clean up resources. Have a look at how the stock plugins do that as an example.

I saw how it is done in the stock plugins, but that is exactly what I was trying to avoid, i.e. doing the same work twice, like creating a function such as `hangup_media_internal()` and having to save states like `session->destroyed` or `session->hangingup`; in other words, avoiding to have to check in `destroy_session` callback or in `hangup_media` itself if it was already called back or not.

> Closing as not an issue.

I'll assume that this is the correct behavior by design, then. Is this topic open for discussion, though? I insist a bit on this because I see this as a misleading behavior that can lead one to bugs and which must be cornered with unneeded complex code, like what is actually done on stock plugins, and because I saw the same pattern in some other places in code.

Anyway, thanks for your attention! :)
